src/main/host/syscall/handler/signal.rs

```diff
@@ -371,11 +370,89 @@ impl SyscallHandler {
     );
     pub fn sigaltstack(
         ctx: &mut SyscallContext,
-        _uss: ForeignPtr<linux_api::signal::stack_t>,
-        _uoss: ForeignPtr<linux_api::signal::stack_t>,
-    ) -> Result<(), SyscallError> {
-        let rv: i32 = Self::legacy_syscall(c::syscallhandler_sigaltstack, ctx)?;
-        assert_eq!(rv, 0);
+        uss: ForeignPtr<linux_api::signal::stack_t>,
+        uoss: ForeignPtr<linux_api::signal::stack_t>,
+    ) -> Result<(), Errno> {
+        log::trace!("sigaltstack({uss:p}, {uoss:p})");
+
+        let shmem_lock = ctx.objs.host.shim_shmem_lock_borrow().unwrap();
+        let thread_shmem = ctx.objs.thread.shmem();
+        let mut thread_protected = thread_shmem.protected.borrow_mut(&shmem_lock.root);
+
+        let old_ss = unsafe { *thread_protected.sigaltstack() };
+
+        if !uss.is_null() {
+            if old_ss.flags_retain().contains(SigAltStackFlags::SS_ONSTACK) {
+                // sigaltstack(2): EPERM An attempt was made to change the
+                // alternate signal stack while it was active.
+                return Err(Errno::EPERM);
+            }
+
+            let mut new_ss = ctx.objs.process.memory_borrow().read(uss)?;
+            if new_ss.flags_retain().contains(SigAltStackFlags::SS_DISABLE) {
+                // sigaltstack(2): To disable an existing stack, specify ss.ss_flags
+                // as SS_DISABLE. In this case, the kernel ignores any other flags
+                // in ss.ss_flags and the remaining fields in ss.
+                new_ss = shadow_pod::zeroed();
+                new_ss.ss_flags = SigAltStackFlags::SS_DISABLE.bits();
+            }
+
+            let unrecognized_flags = new_ss
+                .flags_retain()
+                .difference(SigAltStackFlags::SS_DISABLE | SigAltStackFlags::SS_AUTODISARM);
+
+            if !unrecognized_flags.is_empty() {
+                log::debug!(
+                    "Unrecognized signal stack flags {unrecognized_flags:?} in {:?}",
+                    new_ss.flags_retain(),
+                );
+                // Unrecognized flag.
+                return Err(Errno::EINVAL);
+            }
+
+            *unsafe { thread_protected.sigaltstack_mut() } = new_ss;
+        }
+
+        // TODO: should we move this before we modify the alt stack, so that we don't return EFAULT
```

Comment:
Makes sense to me; ideally with a test for this behavior and to confirm consistency with linux behavior, but not strictly necessary.

Reply:
Thanks for bringing up adding a test. It turns out that my TODO (and following change) was wrong, so I've reverted the change that resolves this TODO. (GitHub doesn't show the original change anymore for some reason.)

I will add a test for the correct behaviour in a follow up MR.